### PR TITLE
Add random access file and test

### PR DESCRIPTION
**Pull request by vnghia**
@mihaimaruseac 
This PR add the first version of `tf_random_access_file` for `gcs` (without caching). This PR contain much boilerplate code so it is bigger than other PR.

### CODE REVIEW

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:38]: At this point it seems you really need a header file for the stuff that gets exposed to testing, since you are doing so many forward declarations duplicating the code.

It's ok if more symbols get exposed from the plugin, TF won't call them directly. It's much better than the decrease in readability and the potential of having the forward declarations here not match the ones in the plugin itself.

Let's make a PR which creates the header file and cleans up these forward declarations and the `STATIC` macro.

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:69]: Can we test that the env variable has this format before continuing?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:69]: I will pass the env to ParseGCSPath

**Comment by vnvo2409** [comment]
@mihaimaruseac 
I add a header to expose symbol. However, this PR contains full of boilerplate code ( absl::string_view for example ). I will make a seperate PR to cleanup the test after this PR is merge

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:67]:
```suggestion
      char* bucket;
      char* object;
      TF_Status* status = TF_NewStatus();
      ParseGCSPath(test_dir, true, &bucket, &object, status);
      free(bucket);
      free(object);
      if (TF_GetCode(status) != TF_OK) {
        TF_DeleteStatus(status);
        return false;
      }
      TF_DeleteStatus(status);
```

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:103]: These 2 seem to be leaking

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc
Comment (mihaimaruseac) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:86]: We should make these strings to automate memory management. Since they don't get exposed across the plugin boundary this is a good change to make.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by vnvo2409** [comment]
@mihaimaruseac 
This PR #40726 add the header to expose function. I think the best is I only keep the code for `tf_random_access_file` so that we could merge 2 PR independently. I will add test on the next PR. If we keep the test in this PR, we will have to wait for #40726 to be merged, after that, we rebase and we have to make some changes as well. What do you think ?

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:103]: No. `tf_random_access_file::Cleanup(file)` will free they since they are copied into `TF_RandomAccessFile* file` by `tf_gcs_filesystem::NewRandomAccessFile(filesystem_, filepath.c_str(), file, status_)`

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:86]: I saw you use `const char*` in posix filesystem. Should I make a PR for posix as well ?

Comment (mihaimaruseac) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:86]: You're right. I didn't convert them to `string` there to not have a conversion from `const char*` to `string` when creating the file (API uses `const char*`) and then again from `string` to `const char*` when displaying the error message. Your choice of whether to convert or not, but in this case you have 2 pointers to keep track of and both are generated by the same function call.

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (mihaimaruseac) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:103]: Oh, you're right.

**Comment by mihaimaruseac** [comment]
Yes, let's wait for #40726

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc
Comment (vnvo2409) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem_test.cc:67]: This will be changed when I change `ParseGCSPath`

**Comment by vnvo2409** [comment]
I find that makes senses to use `std::string` instead of `char*`. After #40726 is merged, I will make another PR to change `ParseGCSPath` and `tf_writable_file` to use `std::string`. Finally, I will rebase and make this PR use `std::string` as well

**Comment by vnvo2409** [comment]
@mihaimaruseac 
This PR use `std::string` instead of `char*` #40776 . I dont use `absl::string_view` because we need `substr` return `std::string` instead of `absl::string_view`

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc
Comment (vnvo2409) [inline tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:95]: I dont know if we should use `absl::SimpleAtoi` or `std::stoll`. `std::stoll` throw an exception when no conversion could be performed. `absl::SimpleAtoi` is more convenient but I really do not want to use it because we have to add a dependecy `absl::strings`

**Comment by vnvo2409** [comment]
@mihaimaruseac Done rebase.

File: tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc
Comment (mihaimaruseac) [reply tensorflow/c/experimental/filesystem/plugins/gcs/gcs_filesystem.cc:95]: Google coding style enforces avoidance of exceptions.

It's ok to depend on `absl::strings`.

**Comment by vnvo2409** [comment]
@mihaimaruseac PR adds all missing function to `tf_writable_file` #40835 . I dont add test since the PR has a lot of code already. That PR will cause a conflict with this PR because I change `Init` of filesystem. Once one of these PRs is merge, I will rebase and edit the other PR

**Comment by vnvo2409** [comment]
@mihaimaruseac Rebase again

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**
